worker/uniter/runner/context/contextfactory.go

```diff
@@ -214,6 +214,16 @@ func (f *contextFactory) HookContext(hookInfo hook.Info) (*HookContext, error) {
 			relation.cache.InvalidateMember(hookInfo.RemoteUnit)
 		}
 		hookName = fmt.Sprintf("%s-%s", relation.Name(), hookInfo.Kind)
+		// Update the private address on the context from the relation settings
```

Comment:
Right, there are 3 places where a charm could ask for "what address do I tell external callers to use to connect":
1. private-address value in relation settings, set up by Juju when the unit enters scope, accessed via relation-get
2. unit-get private-addtess
3. network-get

Before this PR, only the first way was cmr aware. The other 2 always used machine private address. Part of the point of this PR is to have all 3 ways use common, cmr aware code.

If we used a new PrivateAddressForHook attribute on the context, are you saying that's what unit-get private-address would return etc, and if a charm did a relation-get and read the private-address value from there, the result should be either what Juju set initially, or what the charm may have overridden.